Potential race condition

https://github.com/dunglas/frankenphp/actions/runs/5378195414/jobs/9757671795

```
==================
WARNING: DATA RACE
Write at 0x00c002563090 by goroutine 5698:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:357 +0x0
  github.com/dunglas/frankenphp.Shutdown()
      /go/src/app/frankenphp.go:297 +0x44
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  github.com/dunglas/frankenphp_test.testTimeout()
      /go/src/app/frankenphp_test.go:606 +0x18b
  github.com/dunglas/frankenphp_test.TestTimeout_worker()
      /go/src/app/frankenphp_test.go:598 +0x66
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Previous read at 0x00c002563090 by goroutine 5699:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:160 +0x0
  github.com/dunglas/frankenphp.ServeHTTP()
      /go/src/app/frankenphp.go:413 +0x464
  github.com/dunglas/frankenphp.startWorkers.func1()
      /go/src/app/worker.go:67 +0x3d5

Goroutine 5698 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1726 +0xa84
  main.main()
      _testmain.go:139 +0x2e9

Goroutine 5699 (running) created at:
  github.com/dunglas/frankenphp.startWorkers()
      /go/src/app/worker.go:44 +0x344
  github.com/dunglas/frankenphp.Init()
      /go/src/app/frankenphp.go:284 +0x724
  github.com/dunglas/frankenphp_test.runTest()
      /go/src/app/frankenphp_test.go:59 +0x506
  github.com/dunglas/frankenphp_test.testTimeout()
      /go/src/app/frankenphp_test.go:606 +0x18b
  github.com/dunglas/frankenphp_test.TestTimeout_worker()
      /go/src/app/frankenphp_test.go:598 +0x66
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47
==================
    logger.go:130: 2023-06-26T12:41:05.787Z	DEBUG	terminated	{"worker": "/go/src/app/testdata/timeout.php"}
    logger.go:130: 2023-06-26T12:41:05.787Z	DEBUG	terminated	{"worker": "/go/src/app/testdata/timeout.php"}
    logger.go:130: 2023-06-26T12:41:05.789Z	DEBUG	FrankenPHP shut down
    testing.go:1319: race detected during execution of test
--- FAIL: TestTimeout_worker (50.21s)
```

```
==================
WARNING: DATA RACE
Write at 0x00c003a90070 by goroutine 5714:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:357 +0x0
  github.com/dunglas/frankenphp.Shutdown()
      /go/src/app/frankenphp.go:297 +0x44
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  github.com/dunglas/frankenphp_test.testTimeout()
      /go/src/app/frankenphp_test.go:606 +0x18b
  github.com/dunglas/frankenphp_test.TestTimeout_worker()
      /go/src/app/frankenphp_test.go:598 +0x66
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Previous read at 0x00c003a90070 by goroutine 5716:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:160 +0x0
  github.com/dunglas/frankenphp.ServeHTTP()
      /go/src/app/frankenphp.go:413 +0x464
  github.com/dunglas/frankenphp.startWorkers.func1()
      /go/src/app/worker.go:67 +0x3d5

Goroutine 5714 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:2036 +0x8d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.runTests()
      /usr/local/go/src/testing/testing.go:2034 +0x87c
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1906 +0xb44
  main.main()
      _testmain.go:139 +0x2e9

Goroutine 5716 (running) created at:
  github.com/dunglas/frankenphp.startWorkers()
      /go/src/app/worker.go:44 +0x344
  github.com/dunglas/frankenphp.Init()
      /go/src/app/frankenphp.go:284 +0x833
  github.com/dunglas/frankenphp_test.runTest()
      /go/src/app/frankenphp_test.go:59 +0x4ee
  github.com/dunglas/frankenphp_test.testTimeout()
      /go/src/app/frankenphp_test.go:606 +0x18b
  github.com/dunglas/frankenphp_test.TestTimeout_worker()
      /go/src/app/frankenphp_test.go:598 +0x66
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47
==================
    logger.go:130: 2023-06-26T18:05:40.444Z	DEBUG	shutting down	{"worker": "/go/src/app/testdata/timeout.php"}
    logger.go:130: 2023-06-26T18:05:40.452Z	DEBUG	terminated	{"worker": "/go/src/app/testdata/timeout.php"}
    logger.go:130: 2023-06-26T18:05:40.453Z	DEBUG	terminated	{"worker": "/go/src/app/testdata/timeout.php"}
    logger.go:130: 2023-06-26T18:05:40.457Z	DEBUG	FrankenPHP shut down
    testing.go:1446: race detected during execution of test
--- FAIL: TestTimeout_worker (50.31s)
```

--- a/frankenphp.go
+++ b/frankenphp.go
@@ -54,6 +54,7 @@ var (
 	ScriptExecutionError        = errors.New("error during PHP script execution")
 
 	requestChan chan *http.Request
+	done        chan struct{}
 	shutdownWG  sync.WaitGroup
 
 	loggerMu sync.RWMutex
@@ -274,6 +275,7 @@ func Init(options ...Option) error {
 	}
 
 	shutdownWG.Add(1)
+	done = make(chan struct{})
 	requestChan = make(chan *http.Request)
 
 	if C.frankenphp_init(C.int(opt.numThreads)) != 0 {
@@ -295,7 +297,7 @@ func Init(options ...Option) error {
 // Shutdown stops the workers and the PHP runtime.
 func Shutdown() {
 	stopWorkers()
-	close(requestChan)
+	close(done)
 	shutdownWG.Wait()
 	requestChan = nil
 
@@ -410,8 +412,9 @@ func ServeHTTP(responseWriter http.ResponseWriter, request *http.Request) error 
 		rc = v.(chan *http.Request)
 	}
 
-	if rc != nil {
-		rc <- request
+	select {
+	case <-done:
+	case rc <- request:
 		<-fc.done
 	}
 
@@ -420,12 +423,13 @@ func ServeHTTP(responseWriter http.ResponseWriter, request *http.Request) error 
 
 //export go_fetch_request
 func go_fetch_request() C.uintptr_t {
-	r, ok := <-requestChan
-	if !ok {
+	select {
+	case <-done:
 		return 0
-	}
 
-	return C.uintptr_t(cgo.NewHandle(r))
+	case r := <-requestChan:
+		return C.uintptr_t(cgo.NewHandle(r))
+	}
 }
 
 func maybeCloseContext(fc *FrankenPHPContext) {

--- a/frankenphp_test.go
+++ b/frankenphp_test.go
@@ -525,13 +525,10 @@ func testFlush(t *testing.T, opts *testOptions) {
 func TestTimeout_module(t *testing.T) {
 	testTimeout(t, &testOptions{})
 }
-
 func TestTimeout_worker(t *testing.T) {
 	testTimeout(t, &testOptions{workerScript: "timeout.php"})
 }
 func testTimeout(t *testing.T, opts *testOptions) {
-	t.Skip("config-dependant")
-
 	config := frankenphp.Config()
 	if !config.ZendMaxExecutionTimers {
 		t.Skip("Zend Timer is not enabled")

--- a/worker.go
+++ b/worker.go
@@ -108,7 +108,6 @@ func startWorkers(fileName string, nbWorkers int) error {
 func stopWorkers() {
 	workersRequestChans.Range(func(k, v any) bool {
 		workersRequestChans.Delete(k)
-		close(v.(chan *http.Request))
 
 		return true
 	})
@@ -135,12 +134,14 @@ func go_frankenphp_worker_handle_request_start(mrh C.uintptr_t) C.uintptr_t {
 	l := getLogger()
 
 	l.Debug("waiting for request", zap.String("worker", fc.Env["SCRIPT_FILENAME"]))
-	r, ok := <-rc
-	if !ok {
-		// channel closed, server is shutting down
+
+	var r *http.Request
+	select {
+	case <-done:
 		l.Debug("shutting down", zap.String("worker", fc.Env["SCRIPT_FILENAME"]))
 
 		return 0
+	case r = <-rc:
 	}
 
 	fc.currentWorkerRequest = cgo.NewHandle(r)

